### PR TITLE
Add support for action parameters with collection parameters

### PR DESCRIPTION
**Pull request by LuukN2**
Just a little thing I noticed while testing. ODataActionParameters with collection parameters would cause the type in the ClassProperty to be null. This caused an exception when trying to generate the type for the parameters.

I also fixed some misplaced using directives from my previous PR.

### CODE REVIEW

File: src/Common.OData.ApiExplorer/AspNet.OData/ClassProperty.cs [hunk 1]
```diff
@@ -30,7 +30,18 @@ internal ClassProperty( IEnumerable<Assembly> assemblies, IEdmOperationParameter
             Contract.Requires( parameter != null );
 
             Name = parameter.Name;
-            type = parameter.Type.Definition.GetClrType( assemblies );
+
+            if ( parameter.Type.IsCollection() )
+            {
+                var collectionType = parameter.Type.AsCollection();
+                var elementType = collectionType.ElementType().Definition.GetClrType( assemblies );
```
Comment (commonsensesoftware) [inline src/Common.OData.ApiExplorer/AspNet.OData/ClassProperty.cs]: Don't we still have to handle possible substitution here if the element type is an _Entity Type_ or _Complex Type_? The element will be a type reference, but the registered reference _might_ have filtered members in this version, no?

Comment (LuukN2) [reply src/Common.OData.ApiExplorer/AspNet.OData/ClassProperty.cs]: Yeah you're right.

I had to change the interface so a serviceprovider required for making the substitutioncontext is passed. I also added a new test for action parameters with types that should be substituted.

File: src/Common.OData.ApiExplorer/AspNet.OData/ClassProperty.cs [hunk 2]
```diff
@@ -24,13 +24,29 @@ internal ClassProperty( PropertyInfo clrProperty, Type propertyType )
             Attributes = AttributesFromProperty( clrProperty );
         }
 
-        internal ClassProperty( IEnumerable<Assembly> assemblies, IEdmOperationParameter parameter )
+        internal ClassProperty( IServiceProvider services, IEnumerable<Assembly> assemblies, IEdmOperationParameter parameter, IModelTypeBuilder typeBuilder )
         {
```
Comment (commonsensesoftware) [inline src/Common.OData.ApiExplorer/AspNet.OData/ClassProperty.cs]: Need to verify that `services` and `typeBuilder` are not `null` here.

File: src/Common.OData.ApiExplorer/AspNet.OData/DefaultModelTypeBuilder.cs [hunk 3]
```diff
@@ -171,14 +171,14 @@ public Type NewStructuredType( IEdmStructuredType structuredType, Type clrType,
         }
 
         /// <inheritdoc />
-        public Type NewActionParameters( IEdmAction action, ApiVersion apiVersion )
+        public Type NewActionParameters( IServiceProvider services, IEdmAction action, ApiVersion apiVersion )
         {
```
Comment (commonsensesoftware) [inline src/Common.OData.ApiExplorer/AspNet.OData/DefaultModelTypeBuilder.cs]: `services` is not allowed to be `null` here.

File: src/Common.OData.ApiExplorer/AspNet.OData/ClassProperty.cs (hunk 2 above)
Comment (commonsensesoftware) [reply src/Common.OData.ApiExplorer/AspNet.OData/ClassProperty.cs]: Also, do you mind throwing in one more small tweak here? `GetPropertyType` was original a read-only field, then I changed it to a method, but now we don't need the method at all any more. Can you revert the field declaration back to:

```c#
internal readonly Type Type;
```

and dump the `GetPropertyType` method altogether?

File: src/Common.OData.ApiExplorer/AspNet.OData/IModelTypeBuilder.cs [hunk 4]
```diff
@@ -29,14 +29,15 @@ public interface IModelTypeBuilder
         Type NewStructuredType( IEdmStructuredType structuredType, Type clrType, ApiVersion apiVersion );
 
         /// <summary>
-        /// Creates an returns a stronly-typed definition for OData action parameters.
```
Comment (commonsensesoftware) [inline src/Common.OData.ApiExplorer/AspNet.OData/IModelTypeBuilder.cs]: Nice catch. :)

**Comment by commonsensesoftware** [comment]
Nicely done. Rolling it in.

**Pull request merged**